RP2040: (Raspberry Pi Pico W) Add Support for MQTT and the Web Server

skilau opened the pull request:
# What does this implement/fix?

RP2040/Raspberry Pi Pico W: Add Support for MQTT and the Web Server

NOTE: The real magic here is all done by the AsyncTCP_RP2040W library:
 https://github.com/khoih-prog/AsyncTCP_RP2040W
(Which appears to be a substantially modified fork of ESPAsyncTCP)

Everything else here is mostly just glue, connecting up the MQTT and Web server to that ported Async TCP client.

NOTE2:
This patch requires slight modifications to the 'ESPAsyncWebServer" package, and the "async-mqtt-client " package.

The pull request for "async-mqtt-client" is here:
https://github.com/OttoWinter/async-mqtt-client/pull/4

The pull request for "ESPAsyncWebServer" is here:
https://github.com/esphome/ESPAsyncWebServer/pull/22

Once those (hopefully) get approved, the uses of the forked repos in this patch will just go back to the original package names/locations, thus cleaning this patch up a bit.

NOTE3:
I am extremely new to ESPHome, (only about 1.5 weeks of using it), so there was probably a better way to override the platform.io packages to my forked versions, but I didn't know how. 

FINAL NOTE:
The author of the "AsyncTCP_RP2040W", has marked their repo as "Archived".
It works extremely well as far as I can tell, but I do not know the Author, or their reasons to Archive the repo.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [X] RP2040

## Example entry for `config.yaml`:

```yaml

esphome:
  name: picotest
  friendly_name: picotest

rp2040:
  board: rpipicow
  framework:
    # Required until https://github.com/platformio/platform-raspberrypi/pull/36 is merged
    platform_version: https://github.com/maxgerhardt/platform-raspberrypi.git

# Enable logging
logger:

ota:
  password: "<SNIPPED>"

wifi:
  manual_ip:
    static_ip: 192.168.1.58
    gateway: 192.168.1.1
    subnet: 255.255.255.0
  ssid: "<SNIPPED>"
  password: "<SNIPPED>"

  # Enable fallback hotspot in case wifi connection fails
  ap:
    ssid: "picotest Fallback Hotspot"
    password: "<SNIPPED>"

web_server:
  port: 80

mqtt:
  broker: 192.168.1.20
  port: 1883

debug:
  update_interval: 60s

text_sensor:
  - platform: debug
    device:
      name: "Device Info"
    reset_reason:
      name: "Reset Reason"

sensor:
  - platform: debug
    free:
      name: "Heap Free"
    block:
      name: "Heap Max Block"
    loop_time:
      name: "Loop Time"
  - platform: uptime
    name: Uptime Sensor
  - platform: internal_temperature
    name: "Internal Temperature"
  - platform: wifi_signal
    name: "WiFi Signal Sensor"

output:
  - platform: gpio
    pin: 32
    id: led

interval:
  - interval: 1000ms
    then:
      - output.turn_on: led
      - delay: 500ms
      - output.turn_off: led

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

skilau committed:
RP2040: Add Support for MQTT and the Web Server

NOTE: The magic here is all done by the AsyncTCP_RP2040W library:
https://github.com/khoih-prog/AsyncTCP_RP2040W

Everything else here is mostly just glue, connecting up the
MQTT and Web server to that ported Async TCP client.

noahwilliamsson commented:
Nice work.  I've gotten the impression that the maintainer of the async MQTT client hasn't been very active on GitHub for some time so I'm going to tag @HeMan because I noticed that he forked that repo while working on https://github.com/esphome/esphome/pull/4759 (Enable IPv6 for ESP8266 and Raspberry pi pico w (RP2040)).  Perhaps you'll have a higher chance of success by collaborating with him.

HeMan commented:
> Nice work. I've gotten the impression that the maintainer of the async MQTT client hasn't been very active on GitHub for some time so I'm going to tag @HeMan because I noticed that he forked that repo while working on #4759 (Enable IPv6 for ESP8266 and Raspberry pi pico w (RP2040)). Perhaps you'll have a higher chance of success by collaborating with him.

Do a PR against my fork (https://github.com/HeMan/async-mqtt-client) and I'll merge it. I hope my IPv6-PR will go in, otherwise we could create a PR to just change async-mqtt-library.

skilau commented:
> > Nice work. I've gotten the impression that the maintainer of the async MQTT client hasn't been very active on GitHub for some time so I'm going to tag @HeMan because I noticed that he forked that repo while working on #4759 (Enable IPv6 for ESP8266 and Raspberry pi pico w (RP2040)). Perhaps you'll have a higher chance of success by collaborating with him.
> 
> Do a PR against my fork (https://github.com/HeMan/async-mqtt-client) and I'll merge it. I hope my IPv6-PR will go in, otherwise we could create a PR to just change async-mqtt-library.

Done, thank you so much!

skilau committed:
Merge branch 'dev' into RP2040_Add_Support_For_MQTT_And_WebServer

HeMan commented:
> > Do a PR against my fork (https://github.com/HeMan/async-mqtt-client) and I'll merge it. I hope my IPv6-PR will go in, otherwise we could create a PR to just change async-mqtt-library.
> 
> Done, thank you so much!

I've released version 1.1.0 on PlatformIO registry with the changes.

skilau committed:
Merge branch 'dev' into RP2040_Add_Support_For_MQTT_And_WebServer

borpin commented:
Be really good to get this reviewed :)

skilau committed:
Merge branch 'backup' into RP2040_Add_Support_For_MQTT_And_WebServer

skilau committed:
Bump AsyncMqttClient-esphome to 1.1.0, to pick up rp2040 additions.

skilau committed:
Merge branch 'dev' into RP2040_Add_Support_For_MQTT_And_WebServer

skilau committed:
RP2040: Fix set_server() call, was changed/broke during last update

skilau committed:
Merge branch 'dev' into RP2040_Add_Support_For_MQTT_And_WebServer

borpin commented:
Any chance of this being reviewed?

HeMan commented:
> Any chance of this being reviewed?

I think you need to get it to pass all tests first.

skilau committed:
Merge branch 'dev' into RP2040_Add_Support_For_MQTT_And_WebServer

oddstr13 commented:
The failing CI looks to me like it's unrelated to the changes made here?

The Build docker ones are complaining about missing pip3 command, and clang-tidy is complaining about `use of undeclared identifier 'DEBUGV'` in `framework-arduinoespressif8266/libraries/ESP8266WiFi/src/include/UdpContext.h`, neither of which are touched in this PR.

Rebasing on top of dev might help?

---
I'm personally not a fan of merging dev/main/master into a feature branch, but I'm not a maintainer here.
(makes for messy git history, and I think it's easier to make mistakes during merging)

bdraco commented:
The containers are fixed on `dev` so you'll need to merge that in, fix conflicts, and any resulting failures to get a clean test run

skilau committed:
Merge branch 'dev' into RP2040_Add_Support_For_MQTT_And_WebServer

skilau commented:
Well, I did merge against 'dev' just now, made the needed changes, and pushed.
I still get the weird errors in clang-tidy with DEBUGV in a platform (ESP8266WiFi) that I don't touch, so it is kind of unclear what is wrong...

borpin commented:
Is there any chance of this being progressed? I'd love to use this board, but no MQTT support is a major issue.

bdraco commented:
The CI isn't passing here so this can't be reviewed yet.

It should also be split into two or more PRs since its two unique features which will make it easier to review:

- one for MQTT 
- one for Webserver support

LorbusChris commented:
I've opened a rebased and slightly modified version of these changes at https://github.com/esphome/esphome/pull/5998

borpin commented:
> The CI isn't passing here so this can't be reviewed yet.
> 
> It should also be split into two or more PRs since its two unique features which will make it easier to review:
> 
> * one for MQTT
> * one for Webserver support

How about now?

ilium007 commented:
> Still no MQTT support yet, right?

I gave up on ESPHome. Moved to Zigbee mesh.

skilau commented:
> > Still no MQTT support yet, right?
> 
> I gave up on ESPHome. Moved to Zigbee mesh.

I gave up on it as well. 
ESPHome is great, and I use it for a number of ESP32 projects here at home, as well a number of ESPHome Bluetooth Proxies.

However, it was/is very clear that the Pico just wasn't going to get the needed love that was required.
So I have moved on too, and have gone with Zigbee (Ikea devices), or Bluetooth (various other sensors).
